### ISSUE TITLE
[RFC] Unit tests revamp, Possible integration tests

### ISSUE THREAD

**Issue by lukeromanowicz** [DivanteLtd/vue-storefront#2305]
Feature Name: Unit tests revamp
Start Date: 2019-01-30
RFC PR: https://github.com/DivanteLtd/vue-storefront/pull/2414

# Summary:

We need to make a few key decisions:
* What testing tools are we going to use for unit testing?
* Where do we place tests in project structure?
* What is our unit testing strategy?
* What do we do with integration tests? It's more like a yes/no question, we're not getting in details here.

# Motivation

Currently, existing tests (16 of them to be precise) written with mocha & karma are integration tests that might seem like a little bit chaotic. As VSF is growing fast, it's high time to use a more modern, comprehensive tool and start writing actual unit tests. Current unit tests could be moved to integration tests section, or completely removed.

## Unit testing motivation

It gives the following advantages:
* Code documentation - you can easily check scenarios a piece of code can be used, what goes in and what goes out.
* Quick regression check - gives some idea what can possibly go wrong after provided changes. 
* Enforces better code quality - as programmers we're lazy people. Bad code is hard and time consuming to test. Writing unit tests forces better code design.
* Helps spotting bugs early by checking if the code output is right on certain input given.
* Integration with CI could be done in no time.

Drawbacks:
* It requires a little bit of extra time that pays back when noticing errors that otherwise would have gone to master. Unfortunately, in case of beginners in unit testing, it can take a lot of time at first that lowers with experience.

## Integration tests motivation:

Pros:
* automated way to test if all files are getting imported correctly and certain parts/features of application work as they should
* quick to write

Cons:
* not as fast as unit tests (still pretty nimble thanks to not using databases)
* it takes a long initial time to create all the required stubs.

# Solution

As we are talking about writing unit tests from scratch, we can choose whatever testing framework/runner we want. We could go with the same set as it was done before or go for something more modern.

## 1. Available unit testing tools
I decided to include Mocha (current solution) vs Jest (alternative solution). There are more though: https://raygun.com/blog/javascript-unit-testing-frameworks/

### Mocha + additional libraries for mocks and assertions.
* been there for many years and is known by many experienced developers
* build yourself a testing framework
* super flexible

I'd be glad if somebody could write a little bit about the advantages of this set as I have no experience with these particular tools.

### Jest
 [Jest](https://jestjs.io/) is a modern framework by facebook which:
* has large and still a fast-growing community while holding significant market share ([almost 4m downloads](https://www.npmjs.com/package/jest) per week comparing to [mochas 2.3 downloads](https://www.npmjs.com/package/mocha) per week based on npm statistics) 
* test runner included
* is broadly known of being fast, in addition, supports parallel test running out of the box
* has comprehensive mocking possibilities included
* highly configurable, handles *.vue, *.ts and *.js files with ease
* code coverage out of the box

## 2. Test structure
There are two schools of thought regarding tests placement in the project structure.

### Keep tests next to source files
```
module/
└── components/
    ├── Foo.ts
    └── Foo.spec.ts
```

This approach comes in handy when we want to quickly check the component's test, for instance, to check what are the scenarios that the component is made for. The fact of tests existence is hard to be ignored when they are placed next to the source files. In case of doing changes, we can easily jump into test and do appropriate changes in its content. It's good for unit tests but where do we place integration tests regarding this particular module? They could be stored for instance in `module/integration/*` but it feels out of a convention.

### Keep tests in a dedicated directory with a replicated module structure
```
module/
├── components/
│   └── Foo.ts
└── test/
    └── unit/
        └── components/
            └── Foo.spec.ts
```

This approach undeniably gives us flexibility for adding all sorts of tests but they are placed far from the source file. Developers would have to get used to this convention and remember that there are also tests, not only source files, and they shall not be ignored while committing changes to the core. Otherwise, the CI will remind them with a message about not passing unit tests :) 

## 3. Unit testing strategy 

The subject to unit testing is broadly known. Test every component that has some logic in it. There are some black and white examples like always testing helpers or never testing third-party libraries. But what about components that have no logic although they themselves are part of the logic?

I'm referring to our mixins which often play a role of proxy. A great example would be:
https://github.com/DivanteLtd/vue-storefront/blob/6bfe3f1fb6186a8619e51c6019e2d4d2f4854c4d/core/modules/checkout/components/CartSummary.ts#L1-L13

It's the only purpose is to assemble a few other pieces to allow importing all of them with a single mixin. Now the question is: how do we test it?

### Test if using this mixin will export all the things I want to be imported into my target component.
A sample test written in jest would look like:
https://github.com/lukeromanowicz/vue-storefront/blob/9a5a5c2067443b85001cf15767835b308dee0bb5/core/modules/checkout/test/unit/components/CartSummary.spec.ts#L1-L43

In this example, we are sure that certain computed properties and a mixin we need are exported. In theory, everything we wanted to check is covered but it requires a lot of extra code and hacking with checking if the mixin we mocked is really loaded into the final component. It's worth mentioning that the test doesn't look too well and it's hard to read. In other words: that's not what unit tests are designed for, although the objective is achievable.

### Test if a mixin can be initialized (no syntax errors of major mistakes making the mixin unable to be used).  

A sample test, once again written in jest would be:
```
import {shallowMount} from '@vue/test-utils'

import { CartSummary } from '../../../components/CartSummary'

describe('CartSummary', () => {

  it('can be initialized', () => {
    const wrapper = shallowMount({
      template: "<div />",
      mixins: [CartSummary]
    });

    expect(wrapper.exists()).toBe(true);
  })
});
```
Code coverage is still all green but the test checks only basic things like syntax errors or the code being possible to import as a mixin. There are plenty of ways to modify that file and make unusable (i.e. removing all the computed properties and the inner mixin) and the test will still pass.

Prons:
* quick way to automate checking for syntax errors and a possibility to use this component as a mixin
* keeping the code coverage green...

Cons:
* ... which can lead to a false belief that everything is well tested
* require a complimentary test to check if everything is imported correctly (perfect integration test scenario)

### Don't test it at all
Unit testing code coverage for this mixin would be red, but we all know that code coverage is not everything, especially when you use other tools for a complimentary testing.

## 4. Integration tests
We do have E2E Tests which could cover most of the integration tests area if there would be more of them. Unfortunately, these are depended on frontend so we don't have any tool that allows as to test a larger part of VS logic/processes without referring to frontend pages. Also, it takes a lot of time to perform these. One of the solutions could be integration tests.

---
Please share your thoughts as every input is welcome and needed. I'll update this RFC with conclusions as the discussion goes on.

**Comment by janmyszkier** [comment]
Ad4. Integration tests
I'd go with selenium/selenide for integration testing. Is popular, and supports python which is easy to learn even if you're not an experienced tester. (I wouldn't go with java variant because of its learning curve)

For unit I'd go with Jest because if you want people to start writing tests it should be as easy as possible to setup and jest is just that. It's easier to start with it than mocha and has most required things out of the box. I don't think "better" configurability is what we need for this project.

**Comment by lukeromanowicz** [comment]
Personally, I like the idea of utilizing jest to write tests in the second strategy (test only the obvious logic of the component) while covering scenarios of integration with other components with... well... integration tests. Unfortunately, it requires a lot of work to cover everything with not only unit tests but also integration tests. I have a feeling that it would significantly slow down VSF growth as developers who don't write tests (and I recently heard that this is most of us).

Just to be on the same page, let's split testing final combination into e2e and integration tests, I have a feeling that it's been used interchangeably. Integration tests are not running the app in a browser or a browser equivalent - these are e2e tests. Integration tests are like unit tests but focusing on a group of components rather than only one that is isolated from others.

E2E tests are another subject to improve but at least they already exist and check some basic things.

**Comment by janmyszkier** [comment]
>Integration tests are not running the app in a browser or a browser equivalent - these are e2e tests.
ah so you have different integration test in mind. I thought they want to test e.g. Magento + vueSF integration. 
Then again you're right about the "book definition" of the integration tests.

**Comment by lukeromanowicz** [comment]
Unit Tests based on Jest have been merged to develop and will be published in VS 1.9. All old karma tests have been removed.
